core/src/main/java/com/graphhopper/routing/util/parsers/OSMCyclewayParser.java

```diff
@@ -13,81 +13,122 @@
-      cyclewayEnc.setEnum(true, edgeFlags, cycleway);
+    public OSMCyclewayParser(EnumEncodedValue<Cycleway> cyclewayEnc) {
+        this.cyclewayEnc = cyclewayEnc;
+        oneways.add("yes");
+        oneways.add("true");
+        oneways.add("1");
+        oneways.add("-1");
+
+        oppositeLanes.put("opposite", Cycleway.SHARED_LANE);
+        oppositeLanes.put("opposite_lane", Cycleway.LANE);
+        oppositeLanes.put("opposite_track", Cycleway.TRACK);
+        oppositeLanes.put("opposite_share_busway", Cycleway.SHARE_BUSWAY);
     }
-    if (cyclewayForward != Cycleway.MISSING) {
-      cyclewayEnc.setEnum(false, edgeFlags, cyclewayForward);
-      if (isOneway(readerWay)) {
-        cyclewayEnc.setEnum(true, edgeFlags, cyclewayForward);
-      }
+
+    @Override
+    public void createEncodedValues(EncodedValueLookup lookup, List<EncodedValue> list) {
+        list.add(cyclewayEnc);
     }
-    if (cyclewayBackward != Cycleway.MISSING) {
-      cyclewayEnc.setEnum(true, edgeFlags, cyclewayBackward);
-      if (isOneway(readerWay)) {
-        cyclewayEnc.setEnum(false, edgeFlags, cyclewayBackward);
-      }
+
+    @Override
+    public IntsRef handleWayTags(IntsRef edgeFlags, ReaderWay readerWay, IntsRef relationFlags) {
+        // Figuring out what kind of bike infra applies to the forward and backward direction
+        // is complicated.
+        //
+        // The goal here is, given the cycleway, cycleway:both:*, and cycleway:left/right:* tags
+        // and the driving side, to come up with a notion of forward and backward cycleway values,
+        // normalized not to use the "opposite" values.
+        //
+        // To simplify the number of cases, cycleway:both is treated as a shorthand for setting
+        // both cycleway:left and cycleway:right, instead of being treated as a different case.
+        // Note that this is different from how the cyclewag tag is treated (see
+        // https://wiki.openstreetmap.org/wiki/Key:cycleway:both).
+
+        Cycleway cycleway = Cycleway.find(readerWay.getTag("cycleway"));
+
+        // Default forward and backward to the "cycleway" value until we find a more
+        // specific value.
+        Cycleway cyclewayForward = cycleway, cyclewayBackward = cycleway;
+
+        // If we find opposite_* tags, normalize them, and treat that as the final word
+        // (discounting any contradictory left/right tags which would be a tagging error).
+        if (oppositeLanes.containsKey(cycleway)) {
+            cyclewayForward = Cycleway.SHARED_LANE;
+            cyclewayBackward = oppositeLanes.get(cycleway);
+        } else {
+            // Driving side is needed to compute default directionality of cycleway:left/right.
+            // See defaults described in wiki:
+            // https://wiki.openstreetmap.org/wiki/Key:cycleway:right:oneway
+            DrivingSide drivingSide = DrivingSide.find(readerWay.getTag("driving_side"));
```

Comment:
shouldn't this be after the country rule? in the case that the country as a whole drives on the left but this way for some reason drives on the right? could happen in a diverging diamond interchange

https://www.wikiwand.com/en/Diverging_diamond_interchange